lsst-epo/investigations-client
# fix: tokens refresh on save

## What this change does ##

There were several issues with the existing token rotation:
- refreshes were performed during server component rendering, which does not set cookies for the client because [streaming has already begun](https://nextjs.org/docs/app/api-reference/functions/cookies#caveats)
- with tokens not passed on as cookies, a new JWT and refresh token would exist in memory only and quickly be lost, the next refresh would use an invalidated refresh token
- no check or re-check for validity before trying authorized mutations

This is a partial "get-it-working-and-no-more" fix to simply implement token rotation that works when needed. There is additional work to be done, at some point:

- user ID is saved as an unencrypted token and used as the source of identity when performing queries and mutations. Editing the cookie would allow getting/setting other user's answers
- the query client should receive the same authexchange work
- no clean way of getting the user without multiple methods, user is retrieved only from JWT and not validated against the viewer queries.

## Notes for reviewers ##

Frontend only, will add a backend change at a later point to update the token times.

## Testing ##

- before doing anything, you may want to update the JWT settings in the API or project.yaml to be <15min so you aren't waiting around a lot
- log in
- do some answers
- wait around til the JWT is expired
- look at your `craftRefreshToken` cookie before saving 
- save, see that it was successful
- see that the `craftRefreshToken` has changed
